server/src/main/java/org/candlepin/policy/SystemPurposeComplianceRules.java

```diff
@@ -0,0 +1,214 @@
+ * ComplianceRules
+ *
+ * A class used to check consumer compliance status.
+ */
+public class SystemPurposeComplianceRules {
+    private static Logger log = LoggerFactory.getLogger(SystemPurposeComplianceRules.class);
+
+    private EntitlementCurator entCurator;
+    private StatusReasonMessageGenerator generator;
+    private EventSink eventSink;
+    private ConsumerCurator consumerCurator;
+    private ConsumerTypeCurator consumerTypeCurator;
+    private I18n i18n;
+
+    @Inject
+    public SystemPurposeComplianceRules(EntitlementCurator entCurator,
+        StatusReasonMessageGenerator generator, EventSink eventSink, ConsumerCurator consumerCurator,
+        ConsumerTypeCurator consumerTypeCurator, I18n i18n) {
+
+        this.entCurator = entCurator;
+        this.generator = generator;
+        this.eventSink = eventSink;
+        this.consumerCurator = consumerCurator;
+        this.consumerTypeCurator = consumerTypeCurator;
+        this.i18n = i18n;
+    }
+
+    /**
+     * Check compliance status for a consumer on a specific date.
+     *
+     * @param consumer Consumer to check.
+     * @param updateConsumer whether or not to use consumerCurator.update
+     *        (also expensive)
+     * @return Compliance status.
+     */
+    @SuppressWarnings("checkstyle:indentation")
+    public SystemPurposeComplianceStatus getStatus(Consumer consumer,
+        Collection<Entitlement> existingEntitlements, Collection<Entitlement> newEntitlements,
+        boolean updateConsumer, boolean currentCompliance) {
+
+        SystemPurposeComplianceStatus status = new SystemPurposeComplianceStatus(i18n);
+
+        if (existingEntitlements == null) {
+            existingEntitlements = new LinkedList<>();
+        }
+
+        if (newEntitlements == null) {
+            newEntitlements = new LinkedList<>();
+        }
+
+        Set<Entitlement> entitlements = Stream.concat(
+            existingEntitlements.stream(),
+            newEntitlements.stream())
+            .collect(Collectors.toSet());
+
+        for (Entitlement entitlement : entitlements) {
+
+            String unsatisfedRole = consumer.getRole();
+            Set<String> unsatisfiedAddons = new HashSet<>(consumer.getAddOns());
+            String preferredSla = consumer.getServiceLevel();
+            String preferredUsage = consumer.getUsage();
+
+            Set<Product> products = Stream.concat(
+                entitlement.getPool().getProvidedProducts().stream(),
+                entitlement.getPool().getDerivedProvidedProducts().stream()
+            ).collect(Collectors.toSet());
+            products.add(entitlement.getPool().getProduct());
+            Product derivedProduct = entitlement.getPool().getDerivedProduct();
+            if (derivedProduct != null) {
+                products.add(derivedProduct);
+            }
+
+            for (Product product : products) {
+                if (StringUtils.isNotEmpty(unsatisfedRole) && product.hasAttribute("role")) {
```

Comment:
Something important here: The name of the attribute agreed with IT and which I already added a constant for in my current PR [here](https://github.com/candlepin/candlepin/blob/c7dd1d7c7e0777611071ee28dde1adec65125ed4/server/src/main/java/org/candlepin/model/Product.java#L137-L138) is `roles`, not `role`, since a product has many of them.